File: ludka/gravirun_logic.php
<?php
/**
 * Мини-симуляция грави-ритма: движение объектов и проверка столкновений.
 */
function gravirun_step(array $state, float $dt, int $laneCount = 4): array {
    $lanes = range(0, $laneCount-1);
    $state['spawnTimer'] = ($state['spawnTimer'] ?? 0) + $dt;
    $state['spawnDelay'] = $state['spawnDelay'] ?? 0.95;
    if ($state['spawnTimer'] >= $state['spawnDelay']) {
        $state['spawnTimer'] = 0;
        $state['spawnDelay'] = max(0.45, $state['spawnDelay'] * 0.985);
        $roll = mt_rand() / mt_getrandmax();
        $type = $roll < 0.68 ? 'crystal' : ($roll < 0.87 ? 'spike' : 'boost');
        $state['objects'][] = ['lane' => $lanes[array_rand($lanes)], 'y' => -30.0, 'speed' => 170.0, 'type' => $type];
    }

    $state['objects'] = array_map(function($o) use ($dt){
        $o['y'] += $o['speed'] * $dt;
        return $o;
    }, $state['objects'] ?? []);

    $playerLane = $state['playerLane'] ?? 1;
    $playerY = 0.8;
    $score = $state['score'] ?? 0;
    $lives = $state['lives'] ?? 3;
    $boostTimer = $state['boostTimer'] ?? 0.0;

    foreach ($state['objects'] as $idx => $obj) {
        if ($obj['lane'] !== $playerLane) continue;
        if (abs($obj['y'] - $playerY) <= 0.08) {
            if ($obj['type'] === 'crystal') {
                $score += ($boostTimer > 0) ? 2 : 1;
                unset($state['objects'][$idx]);
            } elseif ($obj['type'] === 'spike') {
                if ($boostTimer > 0) {
                    unset($state['objects'][$idx]);
                } else {
                    $lives -= 1;
                    unset($state['objects'][$idx]);
                }
            } else {
                $boostTimer = 1.5;
                unset($state['objects'][$idx]);
            }
        }
    }

    $state['objects'] = array_values(array_filter($state['objects'], fn($o) => $o['y'] < 1.4));
    $state['score'] = $score;
    $state['lives'] = $lives;
    $state['boostTimer'] = max(0.0, $boostTimer - $dt);

    return $state;
}
?>

File: ludka/gravirun_logic.cpp
#include <vector>
#include <random>
#include <algorithm>

struct Object{int lane; double y; double speed; char type;};
struct State{
    std::vector<Object> objects; int playerLane=1; double spawnTimer=0; double spawnDelay=0.95;
    int score=0; int lives=3; double boostTimer=0.0;
};

static std::mt19937 rng{std::random_device{}()};

State gravirun_step(State state, double dt, int laneCount=4){
    state.spawnTimer += dt;
    if(state.spawnTimer >= state.spawnDelay){
        state.spawnTimer = 0;
        state.spawnDelay = std::max(0.45, state.spawnDelay * 0.985);
        std::uniform_real_distribution<double> roll(0.0,1.0);
        double r = roll(rng);
        char t = r < 0.68 ? 'c' : (r < 0.87 ? 's' : 'b');
        std::uniform_int_distribution<int> laneDist(0,laneCount-1);
        state.objects.push_back({laneDist(rng), -30.0, 170.0, t});
    }

    for(auto &o : state.objects){ o.y += o.speed * dt; }

    double playerY = 0.8;
    for(auto &o : state.objects){
        if(o.lane != state.playerLane) continue;
        if(std::abs(o.y - playerY) <= 0.08){
            if(o.type == 'c'){
                state.score += (state.boostTimer > 0) ? 2 : 1;
                o.type = 'x';
            }else if(o.type == 's'){
                if(state.boostTimer > 0){
                    o.type = 'x';
                }else{
                    state.lives -= 1;
                    o.type = 'x';
                }
            }else if(o.type == 'b'){
                state.boostTimer = 1.5;
                o.type = 'x';
            }
        }
    }

    state.objects.erase(std::remove_if(state.objects.begin(), state.objects.end(), [](const Object &o){
        return o.type=='x' || o.y >= 1.4;
    }), state.objects.end());

    state.boostTimer = std::max(0.0, state.boostTimer - dt);
    return state;
}

// Пример: State s; s = gravirun_step(s, 0.016);
